### PR TITLE
Refactor acados Python Interface

### PR DESCRIPTION
**Pull request by FreyJo**
New PR to replace #542 , to see the reduced changes after merge of #538 

I suggest to merge this instead of #540.
Since the Python tests changed in both PRs, I think merging straight this bigger one should be easier.

- [x] Python: remove constraint class
- [x] Python: remove cost class
- [x] Python: implement external cost
- [x] Refactor test with correct pendulum on cart model
- [x] Python naming convention: Classes are CamelCase
- [x] Fix race car example
- [x] Make race car example travis test
- [x] Add minimal MHE example for Python

Note: This PR contains some breaking changes regarding the Python interface, the following should be a draft for corresponding Release Notes:

General:
- Use Python Style conventions, classes are CamelCase
- Use proper class constructor for solvers instead of `generate_*`
- Introduce `AcadosModel`:
   -  class containing all the information to code generate the external CasADi functions that are needed when creating an acados ocp solver or acados integrator.
Thus, this class contains: a) the name of the model
b) all CasADi variables/expressions needed in the CasADi function generation process.
- parameter values are now `parameter_values` within `AcadosOcp`, was `p` within `AcadosOcp.constraints`

Dynamics:
-  `acados_dae() ` -> `AcadosModel()`

Constraints:
- removed the Python class `acados_constraint`, instead the nonlinear constraints that are defined in CasADi symbolics are defined within `AcadosOcpModel`, this simplifies the formulation, as x, u, z, p just have to be provided once.
- The nonlinear constraint `h` within BGH is specified as the field `con_h_expr` within `AcadosModel`.
- The BGP constraint `phi(r)` aka convex over nonlinear constraint is specified using the fields `con_phi_expr`, `con_r_expr`, `con_r_in_phi` within `AcadosOcpModel`, which used to be `con_phi_expr`, `con_r_expr`, `r` within the `acados_constraint` class.

Cost:
- The nonlinear least squares cost term is now specified by setting the CasADi expressions:
`cost_y_expr`, respectively `cost_y_expr_e` within `AcadosOcpModel`, which used to be `expr` within an `acados_cost` instance called `cost_r`, respectively `cost_r_e`.

### CODE REVIEW

File: examples/acados_python/pendulum_example/export_mhe_solver.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    ocp_mhe.solver_options.nlp_solver_type = 'SQP'
+    # ocp_mhe.solver_options.nlp_solver_type = 'SQP_RTI'
+    ocp_mhe.solver_options.nlp_solver_max_iter = 200
+
+    acados_solver_mhe = generate_ocp_solver(ocp_mhe, json_file = 'acados_ocp.json')
+
+    # set arrival cost weighting matrix
+    acados_solver_mhe.cost_set(0, "W", block_diag(R, Q, Q0))
+
+    return acados_solver_mhe
```
Comment (zanellia) [inline examples/acados_python/pendulum_example/export_mhe_solver.py]: really cool ;)

File: examples/acados_python/pendulum_example/export_pendulum_ode_model.py [hunk 2]
```diff
@@ -63,10 +63,10 @@ def export_pendulum_ode_model():
     xdot = vertcat(x1_dot, theta_dot, v1_dot, dtheta_dot)
 
     # algebraic variables
-    z = SX.sym('z', 0)
+    z = None
```
Comment (zanellia) [inline examples/acados_python/pendulum_example/export_pendulum_ode_model.py]: I lost track of the status here. Why can't `None` and `[]` be set in the constructor?

File: examples/acados_python/pendulum_example/export_pendulum_ode_model.py [hunk 3]
```diff
@@ -63,10 +63,10 @@ def export_pendulum_ode_model():
...
 
     # parameters
-    p = SX.sym('parameters', 0)
+    p = []
```
Comment (zanellia) [inline examples/acados_python/pendulum_example/export_pendulum_ode_model.py]: same here.

File: examples/acados_python/pendulum_example/export_pendulum_ode_model.py (hunk 2 above)
Comment (FreyJo) [reply examples/acados_python/pendulum_example/export_pendulum_ode_model.py]: actually both `None` and `[]` work now. Which is meant to be demonstrated in this file.

**Comment by FreyJo** [comment]
Thanks @zanellia for the discussion making me think again about the class holding the symbolic variables and expressions needed to create acados ocp solvers and integrators.
I hope you the class `AcadosModel` makes sense for you as it is.

**Pull request merged**